File: Study Plan/Top Interview 150/383 Ransom Note/RansomNote.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    383. Ransom Note
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 73 ms (Beats 59.17% of users with C#)
    - Memory 45.46 MB (Beats 20.18% of users with C#)
*/
public class Solution
{
    public bool CanConstruct(string ransomNote, string magazine)
    {
        var ransomNoteLetters = ransomNote.ToList();
        var magazineLetters = magazine.ToList();
        var lettersUsed = 0;

        foreach (var letter in ransomNoteLetters)
        {
            if (magazineLetters.Contains(letter))
            {
                magazineLetters.Remove(letter);
                lettersUsed++;
            }
        }

        return ransomNote.Length == lettersUsed;
    }
}

File: Study Plan/Top Interview 150/0383 Ransom Note/RansomNote.java
/*  
    Problem
    --------------------------------------------------------------------------------
    383. Ransom Note
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 10 ms (Beats 59.10% of users with Java)
    - Memory 44.99 MB (Beats 5.76% of users with Java)
*/

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        char[] ransomNoteLetters = ransomNote.toCharArray();
        char[] magazineLetters = magazine.toCharArray();
        Map<Character, Integer> lettersFromMagazine = new HashMap<>();
        // Create a map of letters from the magazine, where the Key is a letter
        // and the Value is the count of occurrences of each letter
        for (char letter : magazineLetters) {
            lettersFromMagazine.put(letter, lettersFromMagazine.getOrDefault(letter, 0) + 1);
        }
        // Check if the ransomNote can be constructed using the letters
        // from the magazine, looking for each letter in the collection
        for (char letter : ransomNoteLetters) {
            int occurrences = lettersFromMagazine.getOrDefault(letter, 0);
            if (occurrences <= 0) {
                // The letter is not available, or has already been used
                return false;
            }
            // One letter has been used to construct the ransomNote, so it needs
            // to be removed the collection of available letters
            lettersFromMagazine.put(letter, occurrences - 1);
        }
        // The ransomNote can be constructed
        return true;
    }
}